File: JavaSolutions/P06/LoveRect.java
/* Write a function to find the rectangular intersection 
   of two given love rectangles. Love rectangles are always
   "straight" and never "diagonal." More rigorously: each side
   is parallel with either the x-axis or the y-axis.
 */

public class LoveRect {

    public static class Rectangle {
	
	// coordinates of bottom left corner
	Integer leftX;
	Integer bottomY;
	
	// dimensions
	Integer width;
	Integer height;
	
	public Rectangle(Integer leftX, Integer bottomY, Integer width, Integer height) {
	    this.leftX = leftX;
	    this.bottomY = bottomY;
	    this.width  = width;
	    this.height = height;
	}
	
	public Rectangle() {}
	
	public String toString() {
	    return String.format("(%d, %d, %d, %d)", leftX, bottomY, width, height);
	}
    }

    public static int[] intervalIntersect(int[] i, int[] j) {
	int beg = i[0] > j[0] ? i[0] : j[0];
	int end = i[1] < j[1] ? i[1] : j[1];
	if (beg < end) return new int[] {beg, end};
	return null;
    }

    public static Rectangle rectIntersect(Rectangle x, Rectangle y) {
	int[] xv = new int[] {x.bottomY, x.bottomY + x.height};
	int[] xh = new int[] {x.leftX, x.leftX + x.width};

	int[] yv = new int[] {y.bottomY, y.bottomY + y.height};
	int[] yh = new int[] {y.leftX, y.leftX + y.width};

	int[] iv = intervalIntersect(xv, yv);
	int[] ih = intervalIntersect(xh, yh);

	if (iv == null || ih == null) return null;

	return new Rectangle(ih[0], iv[0], ih[1] - ih[0], iv[1] - iv[0]);
    }

    public static void test(Rectangle x, Rectangle y) {
	System.out.println("Rectangle 1: " + x.toString());
	System.out.println("Rectangle 2: " + y.toString());
	Rectangle intersect = rectIntersect(x, y);
	System.out.println("Intersect:   " + (intersect != null ? intersect.toString() : "null"));
	System.out.println("______________________________________");
    }

    public static void main(String[] args) {
	Rectangle x, y;

	x = new Rectangle(0, 0, 6, 3);
	y = new Rectangle(4, -4, 4, 6);
	test(x, y);

        x = new Rectangle(0, 3, 5, 5);
	y = new Rectangle(3, 0, 1, 9);
	test(x, y);

	x = new Rectangle(0, 0, 8, 4);
	y = new Rectangle(4, 1, 3, 2);
	test(x, y);

	x = new Rectangle(0, 1, 5, 3);
	y = new Rectangle(5, 0, 2, 5);
	test(x, y);

	x = new Rectangle(0, 8, 8, 6);
	y = new Rectangle(4, 0, 8, 6);
	test(x, y);

    }

}

File: PythonSolutions/P06/LoveRect.py
# Write a function to find the rectangular intersection of two given rectangles.
# Rectangles are defined as dictionaries like this:
#
#  my_rectangle = {
#    'left_x': 1,
#    'bottom_y': 5,
#    'width': 10,
#    'height': 4,
#  }

import math

def rectIntersect(rect1, rect2):
	top1 = (rect1['left_x'], rect1['left_x'] + rect1['width'])
	top2 = (rect2['left_x'], rect2['left_x'] + rect2['width'])
	top = intervalIntersect(top1, top2)

	left1 = (rect1['bottom_y'], rect1['bottom_y'] + rect1['height'])
	left2 = (rect2['bottom_y'], rect2['bottom_y'] + rect2['height'])
	left = intervalIntersect(left1, left2)

	if math.isnan(top[0]) or math.isnan(top[1]) or math.isnan(left[0]) or math.isnan(left[1]):
		top = left = (float('nan'), float('nan'))
	return {
		'left_x' : top[0],
		'bottom_y' : left[0],
		'width' : top[1] - top[0],
		'height' : left[1] - left[0], 
	}


def intervalIntersect(ival1, ival2):
	beg = max(ival1[0], ival2[0])
	end = min(ival1[1], ival2[1])
	if beg <= end:
		return (beg, end)
	else:
		return (float('nan'), float('nan'))

def test(rect1, rect2):
	isect = rectIntersect(rect1, rect2)
	print "(%s, %s, %s, %s)" % (isect['left_x'], isect['bottom_y'], isect['width'], isect['height'])

rect1 = {'left_x' : 0, 'bottom_y' : 5, 'width' : 10, 'height' : 12}
rect2 = {'left_x' : 3, 'bottom_y' : -5, 'width' : 14, 'height' : 12}
test(rect1, rect2)

rect1 = {'left_x' : 0, 'bottom_y' : 5, 'width' : 10, 'height' : 12}
rect2 = {'left_x' : 3, 'bottom_y' : 8, 'width' : 6, 'height' : 7}
test(rect1, rect2)

rect1 = {'left_x' : 0, 'bottom_y' : 5, 'width' : 10, 'height' : 12}
rect2 = {'left_x' : 10, 'bottom_y' : -5, 'width' : 14, 'height' : 10}
test(rect1, rect2)

rect1 = {'left_x' : 0, 'bottom_y' : 5, 'width' : 10, 'height' : 12}
rect2 = {'left_x' : 10, 'bottom_y' : -5, 'width' : 14, 'height' : 12}
test(rect1, rect2)

rect1 = {'left_x' : 0, 'bottom_y' : 5, 'width' : 10, 'height' : 12}
rect2 = {'left_x' : 3, 'bottom_y' : -5, 'width' : 14, 'height' : 8}
test(rect1, rect2)